src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx

```diff
@@ -0,0 +1,131 @@
+  dataSources, // list of all available datasource connections.
+  dataSourceOptionList, // combo box renderable option list derived from dataSources
+  selectedSources, // current selected datasource in the form of [{ label: xxx, value: xxx }]
+  onDataSourceSelect,
+  setDataSourceOptionList,
+  onFetchDataSetError, //   onFetchDataSetError, Callback for handling dataset fetch errors. Ensure it's memoized.
+  singleSelection = true,
+}: DataSourceSelectableProps) => {
+  // This effect fetches datasets and prepares the datasource list for UI rendering.
+  useEffect(() => {
+    // Fetches datasets for a given datasource and returns it along with the source.
+    const fetchDataSetWithSource = async (ds: GenericDataSource): Promise<DataSetWithSource> => {
+      const dataSet = await ds.getDataSet();
+      return {
+        ds,
+        data_sets: dataSet,
+      };
+    };
+
+    // Map through all data sources and fetch their respective datasets.
+    const fetchDataSets = () => dataSources.map((ds: DataSourceType) => fetchDataSetWithSource(ds));
+
+    const isIndexPatterns = (dataset: string | IndexPatternOption): boolean => {
+      if (typeof dataset === 'string') return false;
+
+      return dataset.title !== undefined && dataset.id !== undefined;
+    };
+
+    // Get the option format for the combo box from the dataSource and dataSet.
+    const getSourceOptions = (dataSource: DataSourceType, dataSet: DataSetType) => {
+      const optionContent = {
+        type: dataSource.getType(),
+        name: dataSource.getName(),
+        ds: dataSource,
+      };
+      if (isIndexPatterns(dataSet)) {
+        const ip = dataSet as IndexPatternOption;
+        return {
+          ...optionContent,
+          label: ip.title,
+          value: ip.id,
+        };
+      }
+      return {
+        ...optionContent,
+        label: dataSource.getName(),
+        value: dataSource.getName(),
+      };
+    };
+
+    // Convert fetched datasets into a structured format suitable for selector rendering.
+    const getSourceList = (allDataSets: ISourceDataSet[]) => {
+      const finalList = [] as DataSourceGroup[];
+      allDataSets.forEach((curDataSet) => {
+        const typeKey = curDataSet.ds.getType() as DataSourceTypeKey;
+        const groupName = DATASOURCE_TYPE_DISPLAY_NAME_MAP[typeKey] || 'Default Group';
+
+        const existingGroup = finalList.find((item) => item.label === groupName);
+        const mappedOptions = curDataSet.data_sets?.map((dataSet) =>
+          getSourceOptions(curDataSet.ds, dataSet)
+        );
+
+        // check if add new datasource group or add to existing one
+        if (existingGroup) {
+          const existingOptionIds = new Set(existingGroup.options.map((opt) => opt.label));
+          const nonDuplicateOptions = mappedOptions.filter(
+            (opt) => !existingOptionIds.has(opt.label)
+          );
+          existingGroup.options.push(...nonDuplicateOptions);
```

Comment:
This seems to be unnecessary. `existingGroup` is a local variable inside the loop and this is the last statement.

Reply:
i agree, actually couldn't understand what the code tries to do.. @mengweieric i'll leave this to you

Reply:
discussed with eric and i took another look, `existingGroup` holds a reference of an element in the `finalList` array. It's not adding elements to `finalList` but modifying the existing group options in it

Reply:
While there are more developer friendly ways of doing this, let's add some comments for the next person.